Add initial support for screen orientation locking / unlocking

cdumez opened the pull request:
#### 9a6f00e09dd1714190585288993b1edb9605b708
<pre>
Add initial support for screen orientation locking / unlocking
<a href="https://bugs.webkit.org/show_bug.cgi?id=246302">https://bugs.webkit.org/show_bug.cgi?id=246302</a>

Reviewed by Darin Adler.

Add initial support for screen orientation locking / unlocking:
- <a href="https://w3c.github.io/screen-orientation/#lock-method">https://w3c.github.io/screen-orientation/#lock-method</a>
- <a href="https://w3c.github.io/screen-orientation/#unlock-method">https://w3c.github.io/screen-orientation/#unlock-method</a>

Locking / Unlocking is now functional in layout tests on iOS simulator,
allowing us to run the WPT tests for this feature.

* LayoutTests/TestExpectations:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/active-lock-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/event-before-promise-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/lock-basic-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/lock-unlock-check-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/onchange-event-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/onchange-event-subframe-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/orientation-reading-expected.txt: Added.
Rebaseline tests on iOS now that they are properly running.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/ScreenOrientation.cpp:
(WebCore::ScreenOrientation::lock):
(WebCore::ScreenOrientation::unlock): Deleted.
(WebCore::ScreenOrientation::type const): Deleted.
(WebCore::ScreenOrientation::angle const): Deleted.
(WebCore::ScreenOrientation::visibilityStateChanged): Deleted.
(WebCore::ScreenOrientation::shouldListenForChangeNotification const): Deleted.
(WebCore::ScreenOrientation::screenOrientationDidChange): Deleted.
(WebCore::ScreenOrientation::activeDOMObjectName const): Deleted.
(WebCore::ScreenOrientation::suspend): Deleted.
(WebCore::ScreenOrientation::resume): Deleted.
(WebCore::ScreenOrientation::stop): Deleted.
(WebCore::ScreenOrientation::virtualHasPendingActivity const): Deleted.
(WebCore::ScreenOrientation::eventListenersDidChange): Deleted.
* Source/WebCore/page/ScreenOrientation.h:
* Source/WebCore/page/ScreenOrientationType.h:
(WebCore::screenOrientationIsPortait):
(WebCore::screenOrientationIsLandscape):
* Source/WebCore/platform/ScreenOrientationManager.cpp: Copied from Source/WebCore/platform/ios/Device.h.
(WebCore::ScreenOrientationManager::setLockPromise):
(WebCore::ScreenOrientationManager::takeLockPromise):
* Source/WebCore/platform/ScreenOrientationManager.h:
(WebCore::ScreenOrientationManager::~ScreenOrientationManager): Deleted.
* Source/WebCore/platform/ios/Device.h:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::lockScreenOrientation):
(API::UIClient::unlockScreenOrientation):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/C/WKPageUIClient.h:
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::toWKScreenOrientationType):
(WebKit::UIDelegate::UIClient::lockScreenOrientation):
(WebKit::UIDelegate::UIClient::unlockScreenOrientation):
(WebKit::toWKScreenOrientationLockType): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::didExitFullscreen):
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
(WebKit::WebScreenOrientationManagerProxy::~WebScreenOrientationManagerProxy):
(WebKit::resolveScreenOrientationLockType):
(WebKit::WebScreenOrientationManagerProxy::lock):
(WebKit::WebScreenOrientationManagerProxy::unlock):
(WebKit::WebScreenOrientationManagerProxy::screenOrientationDidChange):
(WebKit::WebScreenOrientationManagerProxy::unlockIfNecessary):
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::lockScreenOrientationCallback):
(WTR::unlockScreenOrientationCallback):
(WTR::TestController::createWebViewWithOptions):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView initWithFrame:configuration:]):
(-[TestRunnerWKWebView setSupportedInterfaceOrientations:]):
(-[TestRunnerWKWebView supportedInterfaceOrientations]):
* Tools/WebKitTestRunner/ios/PlatformWebViewIOS.mm:
(-[PlatformWebViewController supportedInterfaceOrientations]):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:

(WTR::TestController::restorePortraitOrientationIfNeeded):
The previous way of restoring portrait orientation caused a UI rotation animation
which could lead to flakiness on the next test. To address the flakiness I now lock
to portrait and then unlock, which gets us in portrait mode without any animation.

(WTR::TestController::platformResetStateToConsistentValues):
(WTR::TestController::lockScreenOrientation):
(WTR::TestController::unlockScreenOrientation):
(WTR::restorePortraitOrientationIfNeeded): Deleted.
Add screen orientation locking / unlocking to WebKitTestRunner on iOS.

Canonical link: <a href="https://commits.webkit.org/255459@main">https://commits.webkit.org/255459@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dcb47f37aa75282d2babf714b248960439767ea1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92581 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1800 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23168 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102323 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96582 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1798 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30162 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84986 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98486 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98244 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1214 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79079 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28142 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36572 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16756 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34357 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17935 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3785 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38227 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40548 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40136 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37094 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/page/ScreenOrientationType.h:
```diff
@@ -36,6 +36,16 @@ enum class ScreenOrientationType : uint8_t {
     LandscapeSecondary
 };
 
+inline bool screenOrientationIsPortait(ScreenOrientationType type)
```
This function name is OK but it could just be `isPortrait` since the argument type is a strong enough type. You can’t accidentally call it on something that is not a screen orientation.

Also a function like this can and should be marked `constexpr` instead of `inline`.

cdumez replied:
Ok.

darinadler commented on Source/WebKit/UIProcess/API/C/WKPage.cpp:
```diff
@@ -2190,6 +2190,36 @@ void WKPageSetPageUIClient(WKPageRef pageRef, const WKPageUIClientBase* wkClient
+            case WebCore::ScreenOrientationType::PortraitPrimary:
+                break;
+            }
+            return kWKScreenOrientationTypePortraitPrimary;
```
`ASSERT_NOT_REACHED` here?

cdumez replied:
It can be reached right now but OK. I can handle PortraitPrimary in the switch and then use ASSERT_NOT_REACHED() here.

cdumez committed:
Add initial support for screen orientation locking / unlocking
https://bugs.webkit.org/show_bug.cgi?id=246302

Reviewed by Darin Adler.

Add initial support for screen orientation locking / unlocking:
- https://w3c.github.io/screen-orientation/#lock-method
- https://w3c.github.io/screen-orientation/#unlock-method

Locking / Unlocking is now functional in layout tests on iOS simulator,
allowing us to run the WPT tests for this feature.

* LayoutTests/TestExpectations:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/active-lock-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/event-before-promise-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/lock-basic-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/lock-unlock-check-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/onchange-event-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/onchange-event-subframe-expected.txt: Added.
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/screen-orientation/orientation-reading-expected.txt: Added.
Rebaseline tests on iOS now that they are properly running.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/ScreenOrientation.cpp:
(WebCore::ScreenOrientation::lock):
(WebCore::ScreenOrientation::unlock): Deleted.
(WebCore::ScreenOrientation::type const): Deleted.
(WebCore::ScreenOrientation::angle const): Deleted.
(WebCore::ScreenOrientation::visibilityStateChanged): Deleted.
(WebCore::ScreenOrientation::shouldListenForChangeNotification const): Deleted.
(WebCore::ScreenOrientation::screenOrientationDidChange): Deleted.
(WebCore::ScreenOrientation::activeDOMObjectName const): Deleted.
(WebCore::ScreenOrientation::suspend): Deleted.
(WebCore::ScreenOrientation::resume): Deleted.
(WebCore::ScreenOrientation::stop): Deleted.
(WebCore::ScreenOrientation::virtualHasPendingActivity const): Deleted.
(WebCore::ScreenOrientation::eventListenersDidChange): Deleted.
* Source/WebCore/page/ScreenOrientation.h:
* Source/WebCore/page/ScreenOrientationType.h:
(WebCore::screenOrientationIsPortait):
(WebCore::screenOrientationIsLandscape):
* Source/WebCore/platform/ScreenOrientationManager.cpp: Copied from Source/WebCore/platform/ios/Device.h.
(WebCore::ScreenOrientationManager::setLockPromise):
(WebCore::ScreenOrientationManager::takeLockPromise):
* Source/WebCore/platform/ScreenOrientationManager.h:
(WebCore::ScreenOrientationManager::~ScreenOrientationManager): Deleted.
* Source/WebCore/platform/ios/Device.h:
* Source/WebKit/UIProcess/API/APIUIClient.h:
(API::UIClient::lockScreenOrientation):
(API::UIClient::unlockScreenOrientation):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageSetPageUIClient):
* Source/WebKit/UIProcess/API/C/WKPageUIClient.h:
* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.h:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::toWKScreenOrientationType):
(WebKit::UIDelegate::UIClient::lockScreenOrientation):
(WebKit::UIDelegate::UIClient::unlockScreenOrientation):
(WebKit::toWKScreenOrientationLockType): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCommitLoadForFrame):
(WebKit::WebPageProxy::didExitFullscreen):
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
(WebKit::WebScreenOrientationManagerProxy::~WebScreenOrientationManagerProxy):
(WebKit::resolveScreenOrientationLockType):
(WebKit::WebScreenOrientationManagerProxy::lock):
(WebKit::WebScreenOrientationManagerProxy::unlock):
(WebKit::WebScreenOrientationManagerProxy::screenOrientationDidChange):
(WebKit::WebScreenOrientationManagerProxy::unlockIfNecessary):
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::lockScreenOrientationCallback):
(WTR::unlockScreenOrientationCallback):
(WTR::TestController::createWebViewWithOptions):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.h:
* Tools/WebKitTestRunner/cocoa/TestRunnerWKWebView.mm:
(-[TestRunnerWKWebView initWithFrame:configuration:]):
(-[TestRunnerWKWebView setSupportedInterfaceOrientations:]):
(-[TestRunnerWKWebView supportedInterfaceOrientations]):
* Tools/WebKitTestRunner/ios/PlatformWebViewIOS.mm:
(-[PlatformWebViewController supportedInterfaceOrientations]):
* Tools/WebKitTestRunner/ios/TestControllerIOS.mm:

(WTR::TestController::restorePortraitOrientationIfNeeded):
The previous way of restoring portrait orientation caused a UI rotation animation
which could lead to flakiness on the next test. To address the flakiness I now lock
to portrait and then unlock, which gets us in portrait mode without any animation.

(WTR::TestController::platformResetStateToConsistentValues):
(WTR::TestController::lockScreenOrientation):
(WTR::TestController::unlockScreenOrientation):
(WTR::restorePortraitOrientationIfNeeded): Deleted.
Add screen orientation locking / unlocking to WebKitTestRunner on iOS.

Canonical link: https://commits.webkit.org/255459@main
